### PR TITLE
DARE method

### PR DESCRIPTION
**Pull request by MauriceFromm**
## Issue

[Implement DARE](https://github.com/pwhofman/probly/issues/357)

## Motivation and Context

based on [arXiv](https://arxiv.org/abs/2304.04042)

---

## Public API Changes

-   [ ] No Public API changes
-   [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

---

## Checklist

-   [ ] The changes have been tested locally.
-   [ ] Documentation has been updated (if the public API or usage changes).
-   [ ] A entry has been added to [`CHANGELOG.md`](https://github.com/pwhofman/probly/blob/main/CHANGELOG.md) (if relevant for users).
-   [ ] The code follows the project's [style guidelines](https://github.com/pwhofman/probly/blob/main/.github/CONTRIBUTING.md) and passes minimal code style checks.
-   [ ] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

**Comment by pwhofman** [comment]
Thank you for the pull request! From my understanding, the wrapper you have implemented is aligned with the theoretical analysis in Section 4 of the paper, but it is not what they do in practice. In practice, the DARE method is training an ensemble with loss Eq. 1, 2 that is instantiated in Eq. 3. This loss regularization term is then turned on or off depending on whether the loss \mathcal{L} is below a certain threshold or not (see Algorithm 1). The method in probly.method could just call the ensemble function and the loss function can be implemented in probly.train. @Advueu963 already made a start for the loss function in https://github.com/pwhofman/probly/tree/126-add-anti-regularization-loss-function-train-routine, you can have a look there. The switching between with and without regularization could also be done within the loss or if this becomes too messy, you can just incorporate this in the training loop.

**Comment by MauriceFromm** [comment]
Thank you for your review. I put my focus on a `method.batchensemble`-similar approach because i misunderstood the **parallel optimization** paragraph. I've pushed changes:

The `method.dare` now calls `method.ensemble`, the returned _member list_ is then wrapped by `DareModule` to handle the hyperparameters (`lamda_reg`, `threshold`) for anti-regularization.

`train.dare.torch` now has `class Regularizer`, to calculate the regularization loss, and function `dare_loss_step`, to handle turning on/off anti-regularization.

----

How should **parallel optimization** (chapter 3 Implementation) be adressed, is this solvable by giving each member a separate optimizer?

@pwhofman

**Review by pwhofman — CHANGES_REQUESTED** [review]
I think the implementation is still a bit simpler. According to the implementation in the paper (see e.g. top page 5), the members are trained independently. We only need an ensemble (decorated as a DarePredictor, as you have) which is trained based on some loss and this anti-regularization term. Hence, I would keep the DareRegularizer class, which gets a threshold when it's instantiated. Then, at every epoch the regularizer gets the loss as calculated in the train loop and the model params, and the switching logic as defined in dare_loss_step is used to set the regularization to 0 or whatever the value of Eq. 3 in the paper is. The rest, i.e., the torch_dare_wrapper, DareModule, and the forward  of RegularizerDare can be removed.
I would not worry about the parallelization for now. This is something for later versions of probly.

**Comment by MauriceFromm** [comment]
Thank you for the review. I've simplified the anti regularization. The `RegularizerDare` now is a container for the `threshold`, while `dare_loss_step` takes care of the anti-regularization itself, receiving the `model`, `device`, training `loss` and `regularizer`. 
Furthermore, i've removed `method.dare.torch` itself and moved the `method.ensemble` call to `method.dare._common`.
This should now align with your specified requested changes and the paper.

Additionally i've done some training to observe the anti-regularization, which seems to work as expected. The l2-distance of the members increases, while the accuracy also increases.
@pwhofman

**Review by pwhofman — APPROVED** [review]
